sdks/python/apache_beam/testing/synthetic_pipeline.py

```diff
@@ -418,29 +420,32 @@ def get_range_tracker(self, start_position, stop_position):
     return tracker
 
   @staticmethod
-  def random_bytes(length):
+  def random_bytes(length, generator):
     """Return random bytes."""
     return b''.join(
-        (struct.pack('B', random.getrandbits(8)) for _ in range(length)))
+        (struct.pack('B', generator.getrandbits(8)) for _ in range(length)))
 
-  def _gen_kv_pair(self, index):
-    random.seed(index)
-    rand = random.random()
+  def _gen_kv_pair(self, generator, index):
+    generator.seed(index)
+    rand = generator.random()
 
     # Determines whether to generate hot key or not.
     if rand < self._hot_key_fraction:
       # Generate hot key.
       # An integer is randomly selected from the range [0, numHotKeys-1]
       # with equal probability.
-      random.seed(index % self._num_hot_keys)
-    return self.random_bytes(self._key_size), self.random_bytes(
-      self._value_size)
+      generator.seed(index % self._num_hot_keys)
+    return self.random_bytes(self._key_size, generator), self.random_bytes(
+      self._value_size, generator)
 
   def read(self, range_tracker):
     index = range_tracker.start_position()
+    # Get an instance of pseudo-random number generator
+    generator = self._generators[(
```

Comment:
I don't see a difference in performance if we instantiate the generator in `read`, but don't use the cache. Do we need the cache?

Reply:
My bad, it seems the cache isn't really needed. It could be helpful if `read` was called many, many times with the same range tracker, but I don't think this is true. Thanks.